### DISCUSSION TITLE
Reactpy under the hood

### DISCUSSION THREAD

**Discussion by H-G-11** [idom-team/idom#1017] [Question] [answered]
Hello there,

I am starting to experiment with ReactPy, and I have a few questions regarding the limitations.

How does this work exactly under the hood?

From my first understanding, we would have separate routes for each page. Is this the case? And if yes, would we lose the "SPA" aspect of React?

Also, does everything happen server-side and is updated using VDOM? I have some trouble understanding what is exactly sent to the client.

Thank you very much, and sorry for my lack of understanding.

**Comment by Archmonger** [comment, marked as answer]
You can think of `reactpy` as ReactJS server side rendering, except with a Python server. In order to keep things moving, we utilize an asyncio loop.

The VDOM representation of the HTML is retained server-side. When this representation changes, it is sent to the client to be rendered by ReactJS (_technically [Preact](https://preactjs.com)_).

SPA functionality can be obtained by using [`reactpy-router`](https://github.com/reactive-python/reactpy-router).

You can find some more technical details sprinkled around the comment section of our [Reddit post](https://www.reddit.com/r/Python/comments/1274w68/reactpy_build_reactjs_interfaces_in_pure_python/).

**Reply by H-G-11** [reply]
Thanks a lot!
